### DISCUSSION TITLE
Update fails with OperationalError no such column when ForeignKeys are used

### DISCUSSION THREAD

**Discussion by norbusan** [sqlalchemy/sqlalchemy#8468] [Usage Questions] [answered]
Environment:
- sqlalchemy 1.4.40
- sqlite3 backend
- fastapi

I am trying to update a parent in a one-many relationship
```
class Episode(Base):
    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, nullable=True, index=True)
    airdate = Column(Date, nullable=False, index=True)
    comment = Column(String, nullable=True, index=True)
    segments = relationship("Segment")

class Segment(Base):
    id = Column(Integer, primary_key=True, index=True)
    episode_id = Column(Integer, ForeignKey("episode.id"), nullable=False)
    ...some normal keys
```
I fetch the current episode and then try to update it
```
existing_episode = db.query(Episode).filter(Episode.id == id)
existing_episode.update(episode.__dict__)
```
What happens is that I get a strange SQL error about missing column
```
sqlite3.OperationalError) no such column: episode_id
```
which comes from the SQL code also dumped:
```
[SQL: UPDATE episode SET id = episode_id=?, id=?, title=?, airdate=?, comment=? WHERE episode.id = ?]
[parameters: (None, 1, 'asdasd', '2015-05-18', 'NULL', 1)]
```
What I wonder is why the `episode_id` is there and how it came there ... the dict passed in does not contain `episode_id`, when I dump it I see
```
episode dict = >>{'id': 1, 'title': 'asdasd', 'airdate': datetime.date(2015, 5, 18), 'comment': 'NULL', 'segments': None}<<
```
So it seems to be related with the relationship and the ForeignKey, but this is how the docs say one should implement 1-n relationship: https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#one-to-many

I debugged the `update` function in `sqlalchemy/orm/query.py` and see that the query changes after calling
```
upd = upd.values(values)
```
in the following code (print statements added by myself to the sqlalchemy code):
```
        print(f"DEBUG upd 1 = {upd}")
        ppo = update_args.pop("preserve_parameter_order", False)
        print(f"DEBUG ppo = {ppo} values={values}")
        if ppo:
            upd = upd.ordered_values(*values)
        else:
            upd = upd.values(values)
        print(f"DEBUG upd 2 = {upd}")
```
This gives 
```
DEBUG upd 1 = UPDATE episode SET id=:id, title=:title, airdate=:airdate, comment=:comment
DEBUG ppo = False values={'id': 1, 'title': 'asdasd', 'airdate': datetime.date(2015, 5, 18), 'comment': 'NULL', 'segments': None}
DEBUG upd 2 = UPDATE episode SET id = episode_id=:param_1, id=:id, title=:title, airdate=:airdate, comment=:comment
```
in the log. 

I have no idea how the `episode_id` creeped in here considering it is not part of the `values`, and how that somehow strange SQL code came into existence.

Any advice on how to proceed from here?

Thanks

**Comment by zzzeek** [comment, marked as answer]
> ```
> class Episode(Base):
>     id = Column(Integer, primary_key=True, index=True)
>     title = Column(String, nullable=True, index=True)
>     airdate = Column(Date, nullable=False, index=True)
>     comment = Column(String, nullable=True, index=True)
>     segments = relationship("Segment")
> 
> class Segment(Base):
>     id = Column(Integer, primary_key=True, index=True)
>     episode_id = Column(Integer, ForeignKey("episode.id"), nullable=False)
>     ...some normal keys
> ```
> 
> I fetch the current episode and then try to update it
> 
> ```
> existing_episode = db.query(Episode).filter(Episode.id == id)
> existing_episode.update(episode.__dict__)

step one would be try not using `__dict__` on an ORM mapped class, assuming `episode` is an ORM mapped class.  There is an `_sa_instance_state` in there that might very well be getting in the way.  use an explicit dict with the exact names you want, make sure that works.


> which comes from the SQL code also dumped:
> 
> ```
> [SQL: UPDATE episode SET id = episode_id=?, id=?, title=?, airdate=?, comment=? WHERE episode.id = ?]
> [parameters: (None, 1, 'asdasd', '2015-05-18', 'NULL', 1)]

is that an exact paste?  That's invalid SQL syntax


> What I wonder is why the `episode_id` is there and how it came there ... the dict passed in does not contain `episode_id`, when I dump it I see
> 
> ```
> episode dict = >>{'id': 1, 'title': 'asdasd', 'airdate': datetime.date(2015, 5, 18), 'comment': 'NULL', 'segments': None}<<
> ```

if "episode" is a mapped class, the above is wrong, there should be `_sa_instance_state` inside of `__dict__`.

> 
> So it seems to be related with the relationship and the ForeignKey, but this is how the docs say one should implement 1-n relationship: 

can't see that from what you are showing me.


> Any advice on how to proceed from here?

I sense I'm not getting a complete picture here, please provide a runnable example if manually setting up the dictionary is not working

**Reply by norbusan** [reply]
Hi @zzzeek 

thanks for your quick response!

I will look into your suggestions, but there is only one thing I can answer immediately:

> > ```
> > [SQL: UPDATE episode SET id = episode_id=?, id=?, title=?, airdate=?, comment=? WHERE episode.id = ?]
> > [parameters: (None, 1, 'asdasd', '2015-05-18', 'NULL', 1)]
> > ```
> 
> is that an exact paste? That's invalid SQL syntax


Yes, that is sql code generated by sqlalchemy, as I have shown in the lower part where I added print statements to the source of sqlalchemy to debug the update:

The following is an excerpt from sqlalchemy code with DEBUG statements sprinkled in by me:
```
        print(f"DEBUG upd 1 = {upd}")
        ppo = update_args.pop("preserve_parameter_order", False)
        print(f"DEBUG ppo = {ppo} values={values}")
        if ppo:
            upd = upd.ordered_values(*values)
        else:
            upd = upd.values(values)
        print(f"DEBUG upd 2 = {upd}")
```
Which gives the following output, which shows that the call to `upd = upd.values(values)` creates the incorrect sql statement:
```
DEBUG upd 1 = UPDATE episode SET id=:id, title=:title, airdate=:airdate, comment=:comment
DEBUG ppo = False values={'id': 1, 'title': 'asdasd', 'airdate': datetime.date(2015, 5, 18), 'comment': 'NULL', 'segments': None}
DEBUG upd 2 = UPDATE episode SET id = episode_id=:param_1, id=:id, title=:title, airdate=:airdate, comment=:comment
```


I will try to come up with a minimal example, it is in a rather large system so I need to trim it down.

Thanks

Norbert

**Reply by zzzeek** [reply]
Very unusually, SQLite accepts the incorrect syntax, but then doesn't see an "episode_id" column because the table doesn't have one.

The issue is that you have a relationship in your ``__dict__``.   That is, "segments".  The value is None but that doesn't matter, because the key to be UPDATEd for "segments" is the JOIN condition for that relationship; that's its SQL expression:

```py
(Pdb) Episode.segments.__clause_element__()
<sqlalchemy.sql.annotation.AnnotatedBinaryExpression object at 0x7f20ed791d20>
```

so the statement you are instructing SQLAlchemy to create is more plainly this:

```py
(Pdb) stmt = update(Episode.__table__).values({Episode.id == Segment.episode_id: None})
```

that is, you have "x == y" as a key in VALUES.   That gets confused in the compiler, where any normal database would fail on syntax but SQLite seems to ignore the "id = " part:

```
(Pdb) print(stmt)
UPDATE episode SET id = episode_id=:param_1

```

so the answer stays the same, don't use ``__dict__`` from a Python class as a source of a "dictionary view".

**Reply by zzzeek** [reply]
this is the MCVE for this case, the code you have already provided is sufficient

```py
import datetime

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Date
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy import values
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

e = create_engine("sqlite://", echo=True)

Base = declarative_base()


class Episode(Base):
    __tablename__ = "episode"

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String, nullable=True, index=True)
    airdate = Column(Date, nullable=False, index=True)
    comment = Column(String, nullable=True, index=True)
    segments = relationship("Segment")


class Segment(Base):
    __tablename__ = "segment"

    id = Column(Integer, primary_key=True, index=True)
    episode_id = Column(Integer, ForeignKey("episode.id"), nullable=False)


Base.metadata.create_all(e)

s = Session(e)


values = {
    "id": 1,
    "title": "asdasd",
    "airdate": datetime.date(2015, 5, 18),
    "comment": "NULL",
    "segments": None,
}

q = s.query(Episode).filter(Episode.id == 1)
q.update(values)
```

**Reply by norbusan** [reply]
Thanks a lot for your checks.

I confirm that **not** using `__dict__` but explicitly passing key/value hash fixes the issue of malformed SQL statement.

Thanks a lot!!
